### ISSUE TITLE
Can I obtain the fixed map figure on the jupyter notebook? noninteracitve map.

### ISSUE THREAD

**Issue by yun881201** [raphaelquast/EOmaps#112]
Hello. 
EOmaps has good functions that can be applied to my map data analysis.
However, in my case, the interactive figure output is too slow to use if it is on the cell of the jupyter notebook (%matplotlib notebook). 
If I obtain the interactive figure on the new window (%matplotlib qt), the speed is quite good for use.
Actually, I need a just fixed analysis output around the specific country on the jupyter notebook. So can I obtain this kind of noninterative figure output on the jupyter notebook ??

**Comment by raphaelquast** [comment]
Hey, 
I'm not 100% sure I get whats your problem here, but in general I'd recommend using the `ipympl` backend (e.g. `%matplotlib widget`) instead of the older `%matplotlib notebook` which should give you a much better performance (and it works in both classical notebooks and jupyter-lab).

See the [FAQ: jupyter-notebooks](https://eomaps.readthedocs.io/en/latest/FAQ.html#jupyter-notebooks) in the docs on how to get it running!

**Comment by yun881201** [comment]
> Hey, I'm not 100% sure I get whats your problem here, but in general I'd recommend using the `ipympl` backend (e.g. `%matplotlib widget`) instead of the older `%matplotlib notebook` which should give you a much better performance (and it works in both classical notebooks and jupyter-lab).
> 
> See the [FAQ: jupyter-notebooks](https://eomaps.readthedocs.io/en/latest/FAQ.html#jupyter-notebooks) in the docs on how to get it running!

Thank you for your answer. I tried the way you suggest. The figure from this is also too slow to use... but better than  %matplotlib notebook. And I want to obtain noninteractive fixed map figure that can be position-controlled.

**Comment by raphaelquast** [comment]
can you provide a minimal code-example to reproduce your issue?
(I don't experience any performance issues with `%matpltolib widget` on my side)

**Comment by yun881201** [comment]
> can you provide a minimal code-example to reproduce your issue? (I don't experience any performance issues with `%matpltolib widget` on my side)

I just run an example code of EOmap below:

# ----------- create some example-data
lon, lat = np.meshgrid(np.arange(-20, 40, 0.25), np.arange(30, 60, 0.25))
data = pd.DataFrame(
    dict(lon=lon.flat, lat=lat.flat, data_variable=np.sqrt(lon**2 + lat**2).flat)
)
data = data.sample(15000)  # take 15000 random datapoints from the dataset
# ------------------------------------

m = Maps(crs=4326)
m.add_feature.preset.ocean()
m.add_feature.preset.coastline()
m.set_data(
    data=data,  # a pandas-DataFrame holding the data & coordinates
    parameter="data_variable",  # the DataFrame-column you want to plot
    x="lon",  # the name of the DataFrame-column representing the x-coordinates
    y="lat",  # the name of the DataFrame-column representing the y-coordinates
    crs=4326,
)  # the coordinate-system of the x- and y- coordinates
m.set_shape.rectangles(radius=0.125)
m.plot_map()

m.add_colorbar()

c = m.add_compass((0.05, 0.86), scale=7, patch=None)

m.cb.pick.attach.annotate()  # attach a basic pick-annotation (on left-click)
m.add_logo()  # add a logo

**Comment by yun881201** [comment]
My conda list is follow.

# packages in environment at C:\Users\user\anaconda3\envs\specavail1:
#
# Name                    Version                   Build  Channel
antlr-python-runtime      4.7.2           py310h5588dad_1003    conda-forge
argon2-cffi               21.3.0             pyhd3eb1b0_0  
argon2-cffi-bindings      21.2.0          py310h2bbff1b_0  
asttokens                 2.0.5              pyhd3eb1b0_0  
attrs                     22.1.0                   pypi_0    pypi
backcall                  0.2.0              pyhd3eb1b0_0  
beautifulsoup4            4.11.1          py310haa95532_0  
bleach                    4.1.0              pyhd3eb1b0_0  
blosc                     1.21.1               h74325e0_3    conda-forge
boost-cpp                 1.78.0               h9f4b32c_1    conda-forge
branca                    0.5.0              pyhd8ed1ab_0    conda-forge
brotli                    1.0.9                h8ffe710_7    conda-forge
brotli-bin                1.0.9                h8ffe710_7    conda-forge
brotlipy                  0.7.0           py310he2412df_1004    conda-forge
bzip2                     1.0.8                he774522_0  
ca-certificates           2018.10.15           ha4d7672_0    conda-forge/label/gcc7
cairo                     1.16.0            hd694305_1014    conda-forge
cairocffi                 1.3.0              pyhd8ed1ab_0    conda-forge
cairosvg                  2.5.2              pyhd8ed1ab_0    conda-forge
cartopy                   0.21.0          py310h05326cb_0    conda-forge
certifi                   2022.9.14       py310haa95532_0  
cf-units                  3.1.1           py310h2873277_0    conda-forge
cffi                      1.15.1          py310h2bbff1b_0  
cfitsio                   4.1.0                h5a969a9_0    conda-forge
cftime                    1.6.2           py310h9b08ddd_0    conda-forge
charset-normalizer        2.1.1              pyhd8ed1ab_0    conda-forge
click                     8.1.3           py310h5588dad_0    conda-forge
click-plugins             1.1.1                      py_0    conda-forge
cligj                     0.7.2              pyhd8ed1ab_1    conda-forge
cloudpickle               2.2.0              pyhd8ed1ab_0    conda-forge
colorama                  0.4.5           py310haa95532_0  
colorcet                  3.0.0              pyhd8ed1ab_0    conda-forge
contourpy                 1.0.5           py310h232114e_0    conda-forge
cryptography              37.0.4          py310ha857299_0    conda-forge
cssselect2                0.2.1              pyh9f0ad1d_1    conda-forge
curl                      7.83.1               h789b8ee_0    conda-forge
cycler                    0.11.0             pyhd8ed1ab_0    conda-forge
dask-core                 2022.9.1           pyhd8ed1ab_0    conda-forge
datashader                0.14.2             pyh6c4a22f_0    conda-forge
datashape                 0.5.4                      py_1    conda-forge
debugpy                   1.5.1           py310hd77b12b_0  
decorator                 5.1.1              pyhd3eb1b0_0  
defusedxml                0.7.1              pyhd3eb1b0_0  
descartes                 1.1.0                      py_4    conda-forge
docopt                    0.6.2                    pypi_0    pypi
entrypoints               0.4             py310haa95532_0  
eomaps                    4.4.2              pyhd8ed1ab_0    conda-forge
equi7grid                 0.1.0                    pypi_0    pypi
executing                 0.8.3              pyhd3eb1b0_0  
expat                     2.4.9                h1537add_0    conda-forge
fiona                     1.8.21                   pypi_0    pypi
flask                     2.2.2                    pypi_0    pypi
folium                    0.12.1.post1       pyhd8ed1ab_1    conda-forge
font-ttf-dejavu-sans-mono 2.37                 hab24e00_0    conda-forge
font-ttf-inconsolata      3.000                h77eed37_0    conda-forge
font-ttf-source-code-pro  2.038                h77eed37_0    conda-forge
font-ttf-ubuntu           0.83                 hab24e00_0    conda-forge
fontconfig                2.14.0               hce3cb01_0    conda-forge
fonts-conda-ecosystem     1                             0    conda-forge
fonts-conda-forge         1                             0    conda-forge
fonttools                 4.37.3          py310h8d17308_0    conda-forge
freetype                  2.12.1               h546665d_0    conda-forge
freexl                    1.0.6                ha8e266a_0    conda-forge
fsspec                    2022.8.2           pyhd8ed1ab_0    conda-forge
gdal                      3.4.3                    pypi_0    pypi
geographiclib             2.0                      pypi_0    pypi
geojson                   2.5.0                      py_0    conda-forge
geopandas                 0.11.1             pyhd8ed1ab_0    conda-forge
geopandas-base            0.11.1             pyha770c72_0    conda-forge
geos                      0.2.3                    pypi_0    pypi
geotiff                   1.7.1                h714bc5f_3    conda-forge
gettext                   0.19.8.1          ha2e2712_1008    conda-forge
gmaps                     0.9.0                      py_0    conda-forge
hdf4                      4.2.15               h0e5069d_4    conda-forge
hdf5                      1.12.2          nompi_h2a0e4a3_100    conda-forge
icu                       70.1                 h0e60522_0    conda-forge
idna                      3.4                pyhd8ed1ab_0    conda-forge
intel-openmp              2022.1.0          h57928b3_3787    conda-forge
ipykernel                 6.15.2          py310haa95532_0  
ipympl                    0.9.2              pyhd8ed1ab_0    conda-forge
ipython                   8.5.0              pyh08f2357_1    conda-forge
ipython_genutils          0.2.0              pyhd3eb1b0_1  
ipywidgets                8.0.2              pyhd8ed1ab_1    conda-forge
iris                      3.3.0              pyhd8ed1ab_0    conda-forge
iris-sample-data          2.1.0                      py_0    conda-forge/label/gcc7
itsdangerous              2.1.2                    pypi_0    pypi
jedi                      0.18.1          py310haa95532_1  
jinja2                    3.1.2                    pypi_0    pypi
joblib                    1.2.0              pyhd8ed1ab_0    conda-forge
jpeg                      9e                   h8ffe710_2    conda-forge
js2py                     0.71                     pypi_0    pypi
jsonschema                4.4.0           py310haa95532_0  
jupyter-contrib-core      0.4.0                    pypi_0    pypi
jupyter-contrib-nbextensions 0.5.1                    pypi_0    pypi
jupyter-highlight-selected-word 0.2.0                    pypi_0    pypi
jupyter-latex-envs        1.4.6                    pypi_0    pypi
jupyter-nbextensions-configurator 0.5.0                    pypi_0    pypi
jupyter_client            7.3.5           py310haa95532_0  
jupyter_core              4.10.0          py310haa95532_0  
jupyterlab_pygments       0.1.2                      py_0  
jupyterlab_widgets        3.0.3              pyhd8ed1ab_0    conda-forge
kealib                    1.4.15               hdf81f3a_1    conda-forge
kiwisolver                1.4.4           py310h476a331_0    conda-forge
krb5                      1.19.3               h1176d77_0    conda-forge
lcms2                     2.12                 h2a16943_0    conda-forge
lerc                      4.0.0                h63175ca_0    conda-forge
libblas                   3.9.0              16_win64_mkl    conda-forge
libbrotlicommon           1.0.9                h8ffe710_7    conda-forge
libbrotlidec              1.0.9                h8ffe710_7    conda-forge
libbrotlienc              1.0.9                h8ffe710_7    conda-forge
libcblas                  3.9.0              16_win64_mkl    conda-forge
libcurl                   7.83.1               h789b8ee_0    conda-forge
libdeflate                1.13                 h8ffe710_0    conda-forge
libffi                    3.4.2                hd77b12b_4  
libgdal                   3.5.2                hc386656_1    conda-forge
libglib                   2.72.1               h3be07f2_0    conda-forge
libiconv                  1.16                 he774522_0    conda-forge
libkml                    1.3.0             hf2ab4e4_1015    conda-forge
liblapack                 3.9.0              16_win64_mkl    conda-forge
libnetcdf                 4.8.1           nompi_h85765be_104    conda-forge
libpng                    1.6.38               h19919ed_0    conda-forge
libpq                     14.5                 hfcc5ef8_0    conda-forge
librttopo                 1.1.0               h2842628_11    conda-forge
libsodium                 1.0.18               h62dcd97_0  
libspatialindex           1.9.3                h39d44d4_4    conda-forge
libspatialite             5.0.1               ha17912d_18    conda-forge
libsqlite                 3.39.3               hcfcfb64_0    conda-forge
libssh2                   1.10.0               h680486a_3    conda-forge
libtiff                   4.4.0                h92677e6_3    conda-forge
libwebp-base              1.2.4                h8ffe710_0    conda-forge
libxcb                    1.13              hcd874cb_1004    conda-forge
libxml2                   2.9.14               hf5bbc77_4    conda-forge
libzip                    1.9.2                hfed4ece_1    conda-forge
libzlib                   1.2.12               hcfcfb64_3    conda-forge
llvmlite                  0.38.1          py310h2c03ce5_0    conda-forge
locket                    1.0.0              pyhd8ed1ab_0    conda-forge
lxml                      4.9.1                    pypi_0    pypi
lz4-c                     1.9.3                h8ffe710_1    conda-forge
m2w64-gcc-libgfortran     5.3.0                         6    conda-forge
m2w64-gcc-libs            5.3.0                         7    conda-forge
m2w64-gcc-libs-core       5.3.0                         7    conda-forge
m2w64-gmp                 6.1.0                         2    conda-forge
m2w64-libwinpthread-git   5.0.0.4634.697f757               2    conda-forge
mapclassify               2.4.3              pyhd8ed1ab_0    conda-forge
markupsafe                2.1.1           py310h2bbff1b_0  
matplotlib-base           3.6.0           py310h51140c5_0    conda-forge
matplotlib-inline         0.1.6           py310haa95532_0  
mistune                   0.8.4           py310h2bbff1b_1000  
mkl                       2022.1.0           h6a75c08_874    conda-forge
msys2-conda-epoch         20160418                      1    conda-forge
multipledispatch          0.6.0                      py_0    conda-forge
munch                     2.5.0                      py_0    conda-forge
munkres                   1.1.4              pyh9f0ad1d_0    conda-forge
nbclient                  0.5.13          py310haa95532_0  
nbconvert                 6.4.4           py310haa95532_0  
nbformat                  5.5.0           py310haa95532_0  
nest-asyncio              1.5.5           py310haa95532_0  
netcdf4                   1.6.1           nompi_py310h459bb5f_100    conda-forge
networkx                  2.8.6              pyhd8ed1ab_0    conda-forge
notebook                  6.4.12          py310haa95532_0  
numba                     0.55.2          py310h77579ad_0    conda-forge
numpy                     1.22.4+vanilla           pypi_0    pypi
openjpeg                  2.5.0                hc9384bd_1    conda-forge
openssl                   1.1.1q               h2bbff1b_0  
owslib                    0.27.2             pyhd8ed1ab_1    conda-forge
packaging                 21.3               pyhd8ed1ab_0    conda-forge
pandas                    1.4.3                    pypi_0    pypi
pandocfilters             1.5.0              pyhd3eb1b0_0  
param                     1.12.2             pyh6c4a22f_0    conda-forge
parso                     0.8.3              pyhd3eb1b0_0  
partd                     1.3.0              pyhd8ed1ab_0    conda-forge
pcre                      8.45                 h0e60522_0    conda-forge
pickleshare               0.7.5           pyhd3eb1b0_1003  
pillow                    9.2.0           py310h52929f7_2    conda-forge
pip                       22.1.2          py310haa95532_0  
pipwin                    0.5.2                    pypi_0    pypi
pixman                    0.40.0               h8ffe710_0    conda-forge
poppler                   22.04.0              hb57f792_3    conda-forge
poppler-data              0.4.11               hd8ed1ab_0    conda-forge
postgresql                14.5                 h1c22c4f_0    conda-forge
proj                      9.0.1                h1cfcee9_1    conda-forge
prometheus_client         0.14.1          py310haa95532_0  
prompt-toolkit            3.0.20             pyhd3eb1b0_0  
psutil                    5.9.0           py310h2bbff1b_0  
pthread-stubs             0.4               hcd874cb_1001    conda-forge
pure_eval                 0.2.2              pyhd3eb1b0_0  
pycparser                 2.21               pyhd3eb1b0_0  
pyct                      0.4.6                      py_0    conda-forge
pyct-core                 0.4.6                      py_0    conda-forge
pyepsg                    0.4.0                      py_0    conda-forge
pygments                  2.11.2             pyhd3eb1b0_0  
pyjsparser                2.7.1                    pypi_0    pypi
pyopenssl                 22.0.0             pyhd8ed1ab_0    conda-forge
pyparsing                 3.0.9              pyhd8ed1ab_0    conda-forge
pyprind                   2.11.3                   pypi_0    pypi
pyproj                    3.4.0           py310h6377384_0    conda-forge
pyqt5                     5.15.7                   pypi_0    pypi
pyqt5-qt5                 5.15.2                   pypi_0    pypi
pyqt5-sip                 12.11.0                  pypi_0    pypi
pyrsistent                0.18.0          py310h2bbff1b_0  
pyshp                     2.3.1              pyhd8ed1ab_0    conda-forge
pysmartdl                 1.3.4                    pypi_0    pypi
pysocks                   1.7.1              pyh0701188_6    conda-forge
python                    3.10.4               hbb2ffb3_0  
python-dateutil           2.8.2              pyhd8ed1ab_0    conda-forge
python-fastjsonschema     2.16.2          py310haa95532_0  
python-xxhash             3.0.0           py310he2412df_1    conda-forge
python_abi                3.10                    2_cp310    conda-forge
pytileproj                0.0.16                   pypi_0    pypi
pytz                      2022.2.1           pyhd8ed1ab_0    conda-forge
pytz-deprecation-shim     0.1.0.post0              pypi_0    pypi
pywin32                   302             py310h2bbff1b_2  
pywinpty                  2.0.2           py310h5da7b33_0  
pyyaml                    6.0             py310he2412df_4    conda-forge
pyzmq                     23.2.0          py310hd77b12b_0  
requests                  2.28.1             pyhd8ed1ab_1    conda-forge
rtree                     1.0.0           py310h1cbd46b_1    conda-forge
scikit-learn              1.1.2           py310h3a564e9_0    conda-forge
scipy                     1.9.1           py310h578b7cb_0    conda-forge
send2trash                1.8.0              pyhd3eb1b0_1  
setuptools                63.4.1          py310haa95532_0  
shapely                   1.8.2                    pypi_0    pypi
six                       1.16.0             pyh6c4a22f_0    conda-forge
snappy                    1.1.9                h82413e6_1    conda-forge
soupsieve                 2.3.2.post1              pypi_0    pypi
sqlite                    3.39.2               h2bbff1b_0  
stack_data                0.2.0              pyhd3eb1b0_0  
tbb                       2021.5.0             h91493d7_2    conda-forge
terminado                 0.13.1          py310haa95532_0  
testpath                  0.6.0           py310haa95532_0  
threadpoolctl             3.1.0              pyh8a188c0_0    conda-forge
tiledb                    2.11.2               h5689973_0    conda-forge
tinycss2                  1.1.1              pyhd8ed1ab_0    conda-forge
tk                        8.6.12               h2bbff1b_0  
toolz                     0.12.0             pyhd8ed1ab_0    conda-forge
tornado                   6.2             py310h2bbff1b_0  
traitlets                 5.1.1              pyhd3eb1b0_0  
trimesh                   3.15.1                   pypi_0    pypi
typing-extensions         4.3.0           py310haa95532_0  
typing_extensions         4.3.0           py310haa95532_0  
tzdata                    2022.2                   pypi_0    pypi
tzlocal                   4.2                      pypi_0    pypi
ucrt                      10.0.20348.0         h57928b3_0    conda-forge
udunits2                  2.2.28               h892ecd3_0    conda-forge
unicodedata2              14.0.0          py310he2412df_1    conda-forge
urllib3                   1.26.12                  pypi_0    pypi
vc                        14.2                 h21ff451_1  
vs2015_runtime            14.29.30139          h890b9b1_7    conda-forge
wcwidth                   0.2.5              pyhd3eb1b0_0  
webencodings              0.5.1           py310haa95532_1  
werkzeug                  2.2.2                    pypi_0    pypi
wheel                     0.37.1             pyhd3eb1b0_0  
widgetsnbextension        4.0.3              pyhd8ed1ab_0    conda-forge
win_inet_pton             1.1.0           py310h5588dad_4    conda-forge
wincertstore              0.2             py310haa95532_2  
winpty                    0.4.3                         4  
xarray                    2022.6.0           pyhd8ed1ab_1    conda-forge
xerces-c                  3.2.3                h0e60522_5    conda-forge
xmltodict                 0.13.0             pyhd8ed1ab_0    conda-forge
xorg-libxau               1.0.9                hcd874cb_0    conda-forge
xorg-libxdmcp             1.1.3                hcd874cb_0    conda-forge
xxhash                    0.8.0                h8ffe710_3    conda-forge
xyzservices               2022.9.0           pyhd8ed1ab_0    conda-forge
xz                        5.2.6                h8d14728_0    conda-forge
yaml                      0.2.5                h8ffe710_2    conda-forge
zeromq                    4.3.4                hd77b12b_0  
zlib                      1.2.12               h8cc25b3_3  
zstd                      1.5.2                h7755175_4    conda-forge

**Comment by raphaelquast** [comment]
OK, this is really a problem with the jupyter notebook backends which usually require the figure to be rendered twice (... as I understand some javascript subtleties demand that but I have not much experience on that side...)

However, if you just opt for a static output I actually have a code-snippet (that is not yet included in EOmaps) that should give you exactly what you want... (It is intended to be merged at some stage but so far there was not much use for it)

I've uploaded it as a `gist` here:
https://gist.github.com/raphaelquast/cb6f80de5a636bc4df97bd1521a5a6f1

The idea here is to use a non-gui backend (`%matplotlib agg` to render the images and then just directly render the created buffer to the notebook... it should give you the same performance as the `qt` backend.
(also if you want a speedup and you deal with rectangular datasets, just use `m.set_shape.raster` or `m.set_shape.instead of `m.set_shape.rectangles`


Any comments on the usability of this snippet (and if you think it would be helpful as integral part of EOmaps) is highly appreciated!

With the `snapshot` class from the gist applied, a static image can then always be rendered into the notebook via:
```python
%matplotlib agg
m = Maps()
m.add_feature.preset.coastline()
m.snapshot(clear=True)
```

**Comment by yun881201** [comment]
> OK, this is really a problem with the jupyter notebook backends which usually require the figure to be rendered twice (... as I understand some javascript subtleties demand that but I have not much experience on that side...)
> 
> However, if you just opt for a static output I actually have a code-snippet (that is not yet included in EOmaps) that should give you exactly what you want... (It is intended to be merged at some stage but so far there was not much use for it)
> 
> I've uploaded it as a `gist` here: https://gist.github.com/raphaelquast/cb6f80de5a636bc4df97bd1521a5a6f1
> 
> The idea here is to use a non-gui backend (`%matplotlib agg` to render the images and then just directly render the created buffer to the notebook... it should give you the same performance as the `qt` backend. (also if you want a speedup and you deal with rectangular datasets, just use `m.set_shape.raster` or `m.set_shape.shade_raster` instead of `m.set_shape.rectangles`
> 
> Any comments on the usability of this snippet (and if you think it would be helpful as integral part of EOmaps) is highly appreciated!
> 
> With the `snapshot` class from the gist applied, a static image can then always be rendered into the notebook via:
> 
> ```python
> %matplotlib agg
> m = Maps()
> m.add_feature.preset.coastline()
> m.snapshot(clear=True)
> ```

Thank you very much for giving me a good solution.
I will add the snapshot.py to my eomaps Lib folder and test it. 
And then tell you some comments. Thank you!

**Comment by yun881201** [comment]
> OK, this is really a problem with the jupyter notebook backends which usually require the figure to be rendered twice (... as I understand some javascript subtleties demand that but I have not much experience on that side...)
> 
> However, if you just opt for a static output I actually have a code-snippet (that is not yet included in EOmaps) that should give you exactly what you want... (It is intended to be merged at some stage but so far there was not much use for it)
> 
> I've uploaded it as a `gist` here: https://gist.github.com/raphaelquast/cb6f80de5a636bc4df97bd1521a5a6f1
> 
> The idea here is to use a non-gui backend (`%matplotlib agg` to render the images and then just directly render the created buffer to the notebook... it should give you the same performance as the `qt` backend. (also if you want a speedup and you deal with rectangular datasets, just use `m.set_shape.raster` or `m.set_shape.shade_raster` instead of `m.set_shape.rectangles`
> 
> Any comments on the usability of this snippet (and if you think it would be helpful as integral part of EOmaps) is highly appreciated!
> 
> With the `snapshot` class from the gist applied, a static image can then always be rendered into the notebook via:
> 
> ```python
> %matplotlib agg
> m = Maps()
> m.add_feature.preset.coastline()
> m.snapshot(clear=True)
> ```

This is what I wanted, but one more function is needed for fixed figure.
When using cartopy.crs and plt.pcolormesh, I can controll map position by code likt "ax.set_extent([122,131.1,43,33])".
Can I do this when using m.snapshot of EOmaps too?

**Comment by raphaelquast** [comment]
> When using cartopy.crs and plt.pcolormesh, I can controll map position by code likt "ax.set_extent([122,131.1,43,33])".
> Can I do this when using m.snapshot of EOmaps too?

Sure! The snapshot will simply represent the figure as it would be rendered in an interactive backend!
EOmaps uses cartopy under the hood, so it just adds to cartopy and retains all existing functionalities.
The cartopy geo-axis with all its native functionalities is accessible via `m.ax`

<image src=https://user-images.githubusercontent.com/22773387/191794122-c09e2cbd-d226-4d55-b734-008fc7aaa19c.png width=50%>

**Comment by yun881201** [comment]
> > can you provide a minimal code-example to reproduce your issue? (I don't experience any performance issues with `%matpltolib widget` on my side)
> 
> I use the matplotlib-base version 3.6.0. My code is an example code of EOmap below:
> 
> ```python
> # ----------- create some example-data
> lon, lat = np.meshgrid(np.arange(-20, 40, 0.25), np.arange(30, 60, 0.25))
> data = pd.DataFrame(
>     dict(lon=lon.flat, lat=lat.flat, data_variable=np.sqrt(lon**2 + lat**2).flat)
> )
> data = data.sample(15000)  # take 15000 random datapoints from the dataset
> # ------------------------------------
> 
> m = Maps(crs=4326)
> m.add_feature.preset.ocean()
> m.add_feature.preset.coastline()
> m.set_data(
>     data=data,  # a pandas-DataFrame holding the data & coordinates
>     parameter="data_variable",  # the DataFrame-column you want to plot
>     x="lon",  # the name of the DataFrame-column representing the x-coordinates
>     y="lat",  # the name of the DataFrame-column representing the y-coordinates
>     crs=4326,
> )  # the coordinate-system of the x- and y- coordinates
> m.set_shape.rectangles(radius=0.125)
> m.plot_map()
> 
> m.add_colorbar()
> 
> c = m.add_compass((0.05, 0.86), scale=7, patch=None)
> 
> m.cb.pick.attach.annotate()  # attach a basic pick-annotation (on left-click)
> m.add_logo()  # add a logo
> ```

Today, I draw the interative map for my data.
For my data, the interactive map on jupyter notebook is fast enough to use.  FYI: len(my_data) = 104625.

**Comment by raphaelquast** [comment]
@yun881201 ... interesting... so I guess the problem was more on your side then?

**Comment by raphaelquast** [comment]
I'm closing this since the `m.snapshot()` functionality will be included in the next major release (EOmaps v5.0, to be released soon).
Feel free to re-open if something is still unclear!
